code/obj/critter/misc.dm

```diff
@@ -1040,27 +1040,33 @@
 		src.attacking = 1
 		src.visible_message("<span class='combat'><B>The [src.name]</B> starts to envelop [M]!</span>")
 
-		var/lastloc = M.loc
-		SPAWN(6 SECONDS)
-			if (get_dist(src, M) <= 1 && ((M.loc == lastloc)))
-				if(isliving(M))
-					logTheThing("combat", M, null, "was enveloped by [src] (obj) at [log_loc(src)].") // Some logging for instakill critters would be nice (Convair880).
-					M.ghostize()
-
-					if (iscarbon(M))
-						for(var/obj/item/W in M)
-							if (isitem(W))
-								M.u_equip(W)
-								if (W)
-									W.set_loc(M.loc)
-									W.dropped(M)
-									W.layer = initial(W.layer)
-
-				src.visible_message("<span class='combat'><B>The [src.name]</B> completely envelops [M]!</span>")
-				playsound(src.loc, "sound/impact_sounds/Slimy_Hit_4.ogg", 50, 1)
-				qdel(M)
+		SETUP_GENERIC_ACTIONBAR(src, src, 6 SECONDS, .proc/finish_envelope, list(M), 'icons/mob/critter_ui.dmi', "devour_over", \
+					"", INTERRUPT_ACTION | INTERRUPT_MOVE | INTERRUPT_STUNNED | INTERRUPT_ACT)
+		SPAWN(70)
+			src.attacking = 0 //if we're interuppted, we can attack again 7 seconds later
+
+
+	proc/finish_envelope(var/mob/M)
+		if (M && IN_RANGE(src, M, 1))
+			logTheThing("combat", M, null, "was enveloped by [src] (obj) at [log_loc(src)].") // Some logging for instakill critters would be nice (Convair880).
+			for (var/mob/O in AIviewers(src))
+				O.show_message("<span class='combat'><B>[src]</B> completely envelops [M]!</span>", 1)
+			playsound(src, "sound/impact_sounds/Slimy_Hit_4.ogg", 50, 1)
+
+			M.death()
+			M.ghostize()
+			if (iscarbon(M))
+				for (var/obj/item/W in M)
+					if (istype(W,/obj/item))
```

Comment:
I'm pretty sure that check is unnecessary? I hope that check is unnecessary.

Reply:
yeah probably, but it's in /code/datums/abilities/critter/envelope.dm that the mob version used, so I figured better safe than sorry.

Reply:
https://hackmd.io/@goonstation/code#typecheckless-for-loops

Reply:
do you want me to make these changes to the critter ability datum also? I may as well, but it's a little out-of-scope

Reply:
Out of scope of proposed changes. 👍